# Sanity checks & EDA #1

## Abstract

This notebook aims at making a first batch of data explorations on the original dataset and highlight the potential discrepancies.

Various sanity checks will be conducted.

## Original data exploration

Since the original data size is about 67.22 GB, it would be tedious to load the entire dataset in a notebook and run some EDA tools ecosystem (e.g. Pandas profiling) to assess descriptive statistics.

We will then use BigQuery to extract some analytics.

<br>__Caution is taken that we will focus on the sole purpose of our machine learning task: _predicting the taxi fare solely from the data available at pickup time_.
<br>We will then ignore any variable not fulfilling the requirement above or seeming non informative for the task.__

In [1]:
import pandas as pd

pd.options.display.max_columns = 50

In [2]:
%%bigquery
-- Extract a part of the original data from the Chicago taxi trips table

SELECT
  *
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
LIMIT 10

Downloading: 100%|██████████| 10/10 [00:01<00:00,  7.89rows/s]


unique_key  \
0  ba896687c05da1685e51bc455514183ec26d35d2   
1  3533567781a81aa51f99417f3369fbe94ea88947   
2  5544206a64bd1e98703f43c76e2455e4cfaa8140   
3  46e98459db63f62431a47c5f6d2d6f1efb30d259   
4  68075061fa43e8c54fd061d9967653a57af47590   
5  7c154e86606da08cc5302c9cbb0c09f69b24c5a2   
6  096a4052d0c62cd245ce89017d9eb30bdfaee9d4   
7  bacad1aaeabc7f940893f7063d17f1a5c4f00e10   
8  48445431f0cb4aef342209e0586fac06125eb942   
9  e5d30555fb7cbed0a6c20f32ee493a6eb9bfad5c   

                                             taxi_id  \
0  2ff33fa77fef652380380c5128fcb9406473110e8377cf...   
1  79a6bb5d797fc90fd11fc73e20da7b751eac146c0ce52d...   
2  79a6bb5d797fc90fd11fc73e20da7b751eac146c0ce52d...   
3  79a6bb5d797fc90fd11fc73e20da7b751eac146c0ce52d...   
4  79a6bb5d797fc90fd11fc73e20da7b751eac146c0ce52d...   
5  79a6bb5d797fc90fd11fc73e20da7b751eac146c0ce52d...   
6  79a6bb5d797fc90fd11fc73e20da7b751eac146c0ce52d...   
7  79a6bb5d797fc90fd11fc73e20da7b751eac146c0ce52d...   
8  79a6bb5d797fc90fd11fc73e20da7b751eac146c0ce52d...   
9  79a6bb5d797fc90fd11fc73e20da7b751eac146c0ce52d...   

       trip_start_timestamp        trip_end_timestamp  trip_seconds  \
0 2016-05-31 04:00:00+00:00 2016-05-31 04:00:00+00:00            16   
1 2016-05-31 18:15:00+00:00 2016-05-31 18:30:00+00:00          1047   
2 2016-06-01 07:00:00+00:00 2016-06-01 07:15:00+00:00           528   
3 2016-06-01 08:00:00+00:00 2016-06-01 08:15:00+00:00           386   
4 2016-06-01 15:30:00+00:00 2016-06-01 15:45:00+00:00           752   
5 2016-06-01 16:00:00+00:00 2016-06-01 16:15:00+00:00           500   
6 2016-06-01 19:00:00+00:00 2016-06-01 19:15:00+00:00           925   
7 2016-06-02 10:15:00+00:00 2016-06-02 10:15:00+00:00             3   
8 2016-06-02 11:00:00+00:00 2016-06-02 11:00:00+00:00             1   
9 2016-06-02 11:30:00+00:00 2016-06-02 11:30:00+00:00             1   

   trip_miles  pickup_census_tract  dropoff_census_tract  \
0         0.0                  NaN                   NaN   
1         5.8                  NaN                   NaN   
2         1.7                  NaN                   NaN   
3         1.8                  NaN                   NaN   
4         5.0                  NaN                   NaN   
5         3.5                  NaN                   NaN   
6         3.7                  NaN                   NaN   
7         0.0                  NaN                   NaN   
8         0.0                  NaN                   NaN   
9         0.0                  NaN                   NaN   

   pickup_community_area  dropoff_community_area   fare  tips  tolls  extras  \
0                    NaN                     NaN   2.00   0.0    NaN     0.0   
1                    NaN                     NaN   0.01   0.0    NaN     0.0   
2                    NaN                     NaN   7.60   0.0    NaN     2.0   
3                    NaN                     NaN   0.01   0.0    NaN     0.0   
4                    NaN                     NaN   0.01   0.0    NaN     0.0   
5                    NaN                     NaN   9.60   0.0    NaN     0.0   
6                    NaN                     NaN  12.20   0.0    NaN     0.0   
7                    NaN                     NaN   2.00   0.0    NaN     0.0   
8                    NaN                     NaN   2.00   0.0    NaN     0.0   
9                    NaN                     NaN   2.00   0.0    NaN     0.0   

   trip_total payment_type   company  pickup_latitude  pickup_longitude  \
0        2.00         Cash  303 Taxi              NaN               NaN   
1        0.01         Cash  303 Taxi              NaN               NaN   
2        9.60         Cash  303 Taxi              NaN               NaN   
3        0.01         Cash  303 Taxi              NaN               NaN   
4        0.01         Cash  303 Taxi              NaN               NaN   
5        9.60  Credit Card  303 Taxi              NaN               NaN   
6       12.20         Cash  303 Taxi             

### Description of the fields

__unique_key__:
- unique identifier of the row related to a trip

<br>__taxi_id__:
- unique identifier of the taxi cab

<br>__trip_start_timestamp__:
- timestamp of the trip start (rounded to the nearest 15min)

<br>__trip_end_timestamp__:
- timestamp of the trip end (rounded to the nearest 15min)

<br>__trip_seconds__:
- duration of the trip in seconds

<br>__trip_miles__:
- distance of the trip in miles

<br>__pickup_census_tract__:
- unique identifier of the census tract when trip starts
- _for privacy, the census tract is not shown for some trips_

<br>__dropoff_census_tract__:
- unique identifier of the census tract when trip ends
- _for privacy, the census tract is not shown for some trips_

<br>__pickup_community_area__:
- community area where the trip starts

<br>__dropoff_community_area__:
- community area where the trip ends

<br>__fare__:
- fare price of the trip

<br>__tips__:
- tip amount of the trip
- _cash tips generally will not be recorded

<br>__tolls__:
- additional amount to pay for tolls during the trip

<br>__extras__:
- any extra charge for the trip (ex: additional passenger fee, vomit clean-up fee, Illinois Airport Departure Tax, tech fee)

<br>__trip_total__:
- total cost of the trip ( fare, tips, tolls & extras)

<br>__payment_type__:
- type of payment for the trip

<br>__company__:
- taxi cab company

<br>__pickup_latitude__:
- latitude of the center of the pickup census tract or the community area _if the census tract has been hidden for privacy_

<br>__pickup_longitude__:
- longitude of the center of the pickup census tract or the community area _if the census tract has been hidden for privacy_

<br>__pickup_location__:
- location of the center of the pickup census tract or the community area _if the census tract has been hidden for privacy_

<br>__dropoff_latitude__:
- latitude of the center of the dropoff census tract or the community area _if the census tract has been hidden for privacy_

<br>__dropoff_longitude__:
- longitude of the center of the dropoff census tract or the community area _if the census tract has been hidden for privacy_

<br>__dropoff_location__:
- location of the center of the dropoff census tract or the community area _if the census tract has been hidden for privacy_

### 1. unique_key

This field does not have any interest for our machine learning use case.
<br>Let's drop it and exclude it from any exploration.

### 2. taxi_id

This field should not have any importance for our machine learning use case predicting the fares.

Otherwise it would mean that somehow the fares are biased and some drivers may present some irregularities in their driving modes.
<br>Finding such anomalies would result in a specific machine learning use case and is beyond the purpose of our demonstration.
<br>Then let's drop this field and also exclude it from any exploration.

### 3. trip_start_timestamp

This field is an important data available at pickup time.

We will extract various information from it:
- the year of trip start
- the month of trip start
- the day of the week of trip start
- the hour of trip start
- the nearest 15 min of trip start

But first, let's check the original field `trip_start_timestamp`.

In [3]:
%%bigquery
-- Extract basic stats from trip_start_timestamp

SELECT
  MIN(trip_start_timestamp) AS first_trip_start,
  MAX(trip_start_timestamp) AS last_trip_start,
  COUNTIF(trip_start_timestamp IS NULL) AS missing_values
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.21s/rows]


first_trip_start           last_trip_start  missing_values
0 2013-01-01 00:00:00+00:00 2022-06-01 00:00:00+00:00               0

We have trip records from January, 1st, 2013 to September, 1st, 2019.
<br>There are no discrepancies in terms of `trip_start_timestamp`.

In [4]:
%%bigquery
-- Extract additional time information from trip_start_timestamp

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
)
    
SELECT
  EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
  EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
  daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
  EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
  EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames

LIMIT 10

Downloading: 100%|██████████| 10/10 [00:02<00:00,  4.57rows/s]


TripStartYear  TripStartMonth TripStartDay  TripStartHour  TripStartMinute
0           2019               4     Saturday             12                0
1           2019               4     Saturday             12               15
2           2016              12       Monday             20                0
3           2016              12       Monday             20               30
4           2016              12       Monday             20               30
5           2016              12       Monday             21                0
6           2016              12      Tuesday             15               30
7           2016              12      Tuesday             16               30
8           2016              12      Tuesday             20               15
9           2016              12      Tuesday             20               30

We will use the query above as part of __feature engineering__.

### 4. trip_end_timestamp

This field is not known at pickup time.
<br>Let's drop it and exclude it from any exploration.

### 5. trip_seconds

This field is not known at pickup time.
<br>However such field is essential to price the taxi fare as mentioned in: https://checkertaxichicago.com/rates-table/

Let's check first the field `trip_seconds`.

In [5]:
%%bigquery
-- Extract basic stats from trip_seconds

SELECT
  MIN(trip_seconds) AS shortest_trip,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(10)] AS perc10_duration,
  AVG(trip_seconds) AS avg_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(90)] AS perc90_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(95)] AS perc95_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(99)] AS perc99_duration,
  MAX(trip_seconds) AS longest_trip,
  STDDEV(trip_seconds) AS stddev_duration,
  COUNTIF(trip_seconds IS NULL) AS missing_values
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.19s/rows]


shortest_trip  perc10_duration  avg_duration  perc90_duration  \
0              0              164    796.843067             1680   

   perc95_duration  perc99_duration  longest_trip  stddev_duration  \
0             2327             3713         86400      1244.346538   

   missing_values  
0         1300284

This field has a lot of discrepancies: from 0s trips to maximum trip durations of almost 24 hours, which is a bit off.
<br>Furthermore, the field has 1,295,574 missing values, which represents 0.7 % of the total trip records.

Let's apply some filtering.

Based on the fare rates documentation: https://checkertaxichicago.com/rates-table/, let's apply from now on some filters:
- `trip_start_timestamp >= '2016-01-01 00:00:00 UTC'` since such rates are applied since January, 1st, 2016;
- `fare > 3.25` since the base fare is USD 3.25;
- `trip_seconds > 0`;
- `trip_miles / (trip_seconds / 3600) <= 70` since the max speed authorized in Illinois is 70 mph (cf. https://en.wikipedia.org/wiki/Speed_limits_in_the_United_States_by_jurisdiction).

Let's also add two other filters about the pickup & dropoff locations, mentioned via the census tracts:
- `pickup_census_tract IS NOT NULL`;
- `dropoff_census_tract IS NOT NULL`.

In [6]:
%%bigquery
-- Extract basic stats from trip_seconds

SELECT
  MIN(trip_seconds) AS shortest_trip,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(10)] AS perc10_duration,
  AVG(trip_seconds) AS avg_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(90)] AS perc90_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(95)] AS perc95_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(99)] AS perc99_duration,
  MAX(trip_seconds) AS longest_trip,
  STDDEV(trip_seconds) AS stddev_duration,
  COUNT(trip_seconds) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
  AND fare > 3.25
  AND trip_seconds > 0
  AND trip_miles / (trip_seconds / 3600) <= 70
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.20s/rows]


shortest_trip  perc10_duration  avg_duration  perc90_duration  \
0              1              240    791.169265             1740   

   perc95_duration  perc99_duration  longest_trip  stddev_duration  \
0             2460             3808         86400      1022.784562   

   number_trips  
0      63356746

The current filtering keeps __30%__ of the original trip records.

There are still very large trip durations after the current filtering.
<br>It would be safe to remove the last top 1% which might represent some outliers and fix an upper threshold of __3800 sec__ for `trip_seconds`.

Instead of dropping the field `trip_seconds`, let's create some historical aggregated variables from it.
<br>Let's take the assumption that the pickup & dropoff locations and the day of the week & hour of trip start matter.

__Feature #1__
<br>the historical average of `trip_seconds` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #2__
<br>the historical average _over the last week_ of `trip_seconds` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #3__
<br>the historical average _over the last month_ of `trip_seconds` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #4__
<br>the historical average _over the last 3 months_ of `trip_seconds` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

Features #2, #3 & #4 have the advantage to possibly wrap up some more recent & temporary external factors _(like roadworks)_ which may influence the traffic.

In [7]:
%%bigquery
-- Engineer some additional features from trip_seconds

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
),
    
buffer1 AS (
  SELECT
    trip_start_timestamp,
    EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
    EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
    daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
    EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
    EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute,
    trip_seconds,
    DATE_DIFF(PARSE_DATE('%Y-%m-%d',  FORMAT_TIMESTAMP('%Y-%m-%d', trip_start_timestamp)), '2016-01-01', DAY) refDate,
    pickup_census_tract,
    dropoff_census_tract
  FROM
    `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames
  WHERE
    trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    AND pickup_census_tract IS NOT NULL
    AND dropoff_census_tract IS NOT NULL
    AND fare > 3.25
    AND trip_seconds > 0 AND trip_seconds < 3800
    AND trip_miles / (trip_seconds / 3600) <= 70
)

SELECT
  * EXCEPT (refDate),
  AVG(trip_seconds)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDuration
FROM
  buffer1

LIMIT 10

Downloading: 100%|██████████| 10/10 [00:01<00:00,  8.11rows/s]


trip_start_timestamp  TripStartYear  TripStartMonth TripStartDay  \
0 2017-04-04 13:00:00+00:00           2017               4      Tuesday   
1 2022-01-19 11:45:00+00:00           2022               1    Wednesday   
2 2017-10-27 10:00:00+00:00           2017              10       Friday   
3 2017-08-18 11:30:00+00:00           2017               8       Friday   
4 2017-02-01 10:45:00+00:00           2017               2    Wednesday   
5 2017-03-20 10:30:00+00:00           2017               3       Monday   
6 2016-01-01 02:00:00+00:00           2016               1       Friday   
7 2016-01-01 02:45:00+00:00           2016               1       Friday   
8 2016-08-15 22:15:00+00:00           2016               8       Monday   
9 2016-06-26 03:30:00+00:00           2016               6       Sunday   

   TripStartHour  TripStartMinute  trip_seconds  pickup_census_tract  \
0             13                0          2274          17031010100   
1             11               45          1378          17031010202   
2             10                0           845          17031010300   
3             11               30          1452          17031010300   
4             10               45          1860          17031010501   
5             10               30          1080          17031010502   
6              2                0           960          17031010503   
7              2               45           560          17031010503   
8             22               15           840          17031010702   
9              3               30          1119          17031010702   

   dropoff_census_tract  historical_tripDuration  histOneWeek_tripDuration  \
0           17031980000                      NaN                       NaN   
1           17031320400                      NaN                       NaN   
2           17031081202                      NaN                       NaN   
3           17031320100                      NaN                       NaN   
4           17031980000                      NaN                       NaN   
5           17031070200                      NaN                       NaN   
6           17031020702                      NaN                       NaN   
7           17031020702                    960.0                       NaN   
8           17031020602                      NaN                       NaN   
9           17031160400                      NaN                       NaN   

   histOneMonth_tripDuration  histThreeMonth_tripDuration  
0                        NaN                          NaN  
1                        NaN                          NaN  
2                        NaN                          NaN  
3                        NaN                          NaN  
4                        NaN                          NaN  
5                        NaN                          NaN  
6                        NaN                          NaN  
7                        NaN                          NaN  
8                        NaN                          NaN  
9                        NaN                          NaN

We will use the query above as part of __feature engineering__.

Let's now check the hypothesis that the features #1, #2, #3 & #4 engineered above are informative about the original field `trip_seconds`.

In [8]:
%%bigquery
-- Compute the correlations between trip_seconds and historical_tripDuration,
-- histOneWeek_tripDuration, histOneMonth_tripDuration & histThreeMonth_tripDuration

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
),
    
buffer1 AS (
  SELECT
    trip_start_timestamp,
    EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
    EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
    daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
    EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
    EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute,
    trip_seconds,
    DATE_DIFF(PARSE_DATE('%Y-%m-%d',  FORMAT_TIMESTAMP('%Y-%m-%d', trip_start_timestamp)), '2016-01-01', DAY) refDate,
    pickup_census_tract,
    dropoff_census_tract
  FROM
    `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames
  WHERE
    trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    AND pickup_census_tract IS NOT NULL
    AND dropoff_census_tract IS NOT NULL
    AND fare > 3.25
    AND trip_seconds > 0 AND trip_seconds < 3800
    AND trip_miles / (trip_seconds / 3600) <= 70
),

buffer2 AS (

SELECT
  * EXCEPT (refDate),
  AVG(trip_seconds)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDuration
FROM
  buffer1
)

SELECT
  CORR(trip_seconds, historical_tripDuration) AS histCorr,
  CORR(trip_seconds, histOneWeek_tripDuration) AS hist1WCorr,
  CORR(trip_seconds, histOneMonth_tripDuration) AS hist1MCorr,
  CORR(trip_seconds, histThreeMonth_tripDuration) AS hist3MCorr
FROM
  buffer2

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.30s/rows]


histCorr  hist1WCorr  hist1MCorr  hist3MCorr
0  0.899512    0.875111    0.892802    0.896119

The high correlations between the original field `trip_seconds` & the four features engineered from historical data `historical_tripDuration`, `histOneWeek_tripDuration`, `histOneMonth_tripDuration` & `histThreeMonth_tripDuration` check the hypothesis and the usage of such features for the modelling part.

### 5. trip_miles

This field is not known at pickup time.
<br>However such field is essential to price the taxi fare as mentioned in: https://checkertaxichicago.com/rates-table/

Let's check first the field `trip_miles`.

In [9]:
%%bigquery
-- Extract basic stats from trip_miles

SELECT
  MIN(trip_miles) AS shortest_trip,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(10)] AS perc10_distance,
  AVG(trip_miles) AS avg_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(90)] AS perc90_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(95)] AS perc95_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(99)] AS perc99_distance,
  MAX(trip_miles) AS longest_trip,
  STDDEV(trip_miles) AS stddev_distance,
  COUNTIF(trip_miles IS NULL) AS missing_values
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.27s/rows]


shortest_trip  perc10_distance  avg_distance  perc90_distance  \
0            0.0              0.0      3.328411             11.1   

   perc95_distance  perc99_distance  longest_trip  stddev_distance  \
0            17.17             20.6        3460.0         11.19925   

   missing_values  
0            2861

This field has a lot of discrepancies: from 0 miles trips to maximum trip distances of 3460 miles, which is a bit off.
<br>Furthermore, the field has 2,113 missing values, which can be negligible.

Let's apply some filtering.

Based on the fare rates documentation: https://checkertaxichicago.com/rates-table/, let's apply some filters as previously:
- `trip_start_timestamp >= '2016-01-01 00:00:00 UTC'` since such rates are applied since January, 1st, 2016;
- `fare > 3.25` since the base fare is USD 3.25;
- `trip_seconds > 0`;
- `trip_seconds < 3800`;
- `trip_miles > 0`;
- `trip_miles / (trip_seconds / 3600) <= 70` since the max speed authorized in Illinois is 70 mph (cf. https://en.wikipedia.org/wiki/Speed_limits_in_the_United_States_by_jurisdiction);
- `pickup_census_tract IS NOT NULL`;
- `dropoff_census_tract IS NOT NULL`.

In [10]:
%%bigquery
-- Extract basic stats from trip_miles

SELECT
  MIN(trip_miles) AS shortest_trip,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(10)] AS perc10_distance,
  AVG(trip_miles) AS avg_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(90)] AS perc90_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(95)] AS perc95_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(99)] AS perc99_distance,
  MAX(trip_miles) AS longest_trip,
  STDDEV(trip_miles) AS stddev_distance,
  COUNT(trip_miles) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
  AND fare > 3.25
  AND trip_seconds > 0 AND trip_seconds < 3800
  AND trip_miles > 0
  AND trip_miles / (trip_seconds / 3600) <= 70
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.32s/rows]


shortest_trip  perc10_distance  avg_distance  perc90_distance  \
0           0.01              0.5      3.307697             12.7   

   perc95_distance  perc99_distance  longest_trip  stddev_distance  \
0             17.6            18.72          71.6         5.059756   

   number_trips  
0      55840046

Applying the filters above seem to solve the discrepancies in the distance.
<br>A maximum distance value of 72 miles is acceptable.

Instead of dropping the field `trip_miles`, let's create some historical aggregated variables from it.
<br>Let's take the assumption that the pickup & dropoff locations and the day of the week & hour of trip start matter.

__Feature #1__
<br>the historical average of `trip_miles` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #2__
<br>the historical average _over the last week_ of `trip_miles` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #3__
<br>the historical average _over the last month_ of `trip_miles` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #4__
<br>the historical average _over the last 3 months_ of `trip_miles` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

Features #2, #3 & #4 have the advantage to possibly wrap up some more recent & temporary external factors _(like roadworks)_ which may influence the traffic.

In [11]:
%%bigquery
-- Engineer some additional features from trip_miles

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
),
    
buffer1 AS (
  SELECT
    trip_start_timestamp,
    EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
    EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
    daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
    EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
    EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute,
    trip_miles,
    DATE_DIFF(PARSE_DATE('%Y-%m-%d',  FORMAT_TIMESTAMP('%Y-%m-%d', trip_start_timestamp)), '2016-01-01', DAY) refDate,
    pickup_census_tract,
    dropoff_census_tract
  FROM
    `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames
  WHERE
    trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    AND pickup_census_tract IS NOT NULL
    AND dropoff_census_tract IS NOT NULL
    AND fare > 3.25
    AND trip_seconds > 0 AND trip_seconds < 3800
    AND trip_miles > 0
    AND trip_miles / (trip_seconds / 3600) <= 70
)

SELECT
  * EXCEPT (refDate),
  AVG(trip_miles)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDistance
FROM
  buffer1

LIMIT 10

Downloading: 100%|██████████| 10/10 [00:01<00:00,  8.38rows/s]


trip_start_timestamp  TripStartYear  TripStartMonth TripStartDay  \
0 2017-04-22 17:45:00+00:00           2017               4     Saturday   
1 2016-02-23 14:15:00+00:00           2016               2      Tuesday   
2 2017-01-20 11:15:00+00:00           2017               1       Friday   
3 2016-01-12 08:30:00+00:00           2016               1      Tuesday   
4 2016-07-30 22:45:00+00:00           2016               7     Saturday   
5 2017-01-03 11:00:00+00:00           2017               1      Tuesday   
6 2016-12-24 15:00:00+00:00           2016              12     Saturday   
7 2019-07-05 15:00:00+00:00           2019               7       Friday   
8 2019-01-10 22:15:00+00:00           2019               1     Thursday   
9 2019-01-10 22:15:00+00:00           2019               1     Thursday   

   TripStartHour  TripStartMinute  trip_miles  pickup_census_tract  \
0             17               45       10.00          17031010202   
1             14               15       11.60          17031010202   
2             11               15        3.10          17031010300   
3              8               30        0.40          17031010300   
4             22               45       12.30          17031010300   
5             11                0        1.70          17031010300   
6             15                0       20.00          17031010400   
7             15                0       12.63          17031010503   
8             22               15        0.04          17031020400   
9             22               15        0.22          17031020400   

   dropoff_census_tract  historical_tripDistance  histOneWeek_tripDistance  \
0           17031081500                      NaN                       NaN   
1           17031839100                      NaN                       NaN   
2           17031030706                      NaN                       NaN   
3           17031070102                      NaN                       NaN   
4           17031221300                      NaN                       NaN   
5           17031810200                      NaN                       NaN   
6           17031710800                      NaN                       NaN   
7           17031838200                      NaN                       NaN   
8           17031020400                      NaN                       NaN   
9           17031020400                     0.04                       NaN   

   histOneMonth_tripDistance  histThreeMonth_tripDistance  
0                        NaN                          NaN  
1                        NaN                          NaN  
2                        NaN                          NaN  
3                        NaN                          NaN  
4                        NaN                          NaN  
5                        NaN                          NaN  
6                        NaN                          NaN  
7                        NaN                          NaN  
8                        NaN                          NaN  
9                        NaN                          NaN

We will use the query above as part of __feature engineering__.

Let's now check the hypothesis that the features #1, #2, #3 & #4 engineered above are informative about the original field `trip_miles`.

In [12]:
%%bigquery
-- Compute the correlations between trip_miles and historical_tripDistance,
-- histOneWeek_tripDistance, histOneMonth_tripDistance & histThreeMonth_tripDistance

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
),
    
buffer1 AS (
  SELECT
    trip_start_timestamp,
    EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
    EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
    daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
    EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
    EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute,
    trip_miles,
    DATE_DIFF(PARSE_DATE('%Y-%m-%d',  FORMAT_TIMESTAMP('%Y-%m-%d', trip_start_timestamp)), '2016-01-01', DAY) refDate,
    pickup_census_tract,
    dropoff_census_tract
  FROM
    `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames
  WHERE
    trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    AND pickup_census_tract IS NOT NULL
    AND dropoff_census_tract IS NOT NULL
    AND fare > 3.25
    AND trip_seconds > 0 AND trip_seconds < 3800
    AND trip_miles > 0
    AND trip_miles / (trip_seconds / 3600) <= 70
),

buffer2 AS (

SELECT
  * EXCEPT (refDate),
  AVG(trip_miles)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDistance
FROM
  buffer1
)

SELECT
  CORR(trip_miles, historical_tripDistance) AS histCorr,
  CORR(trip_miles, histOneWeek_tripDistance) AS hist1WCorr,
  CORR(trip_miles, histOneMonth_tripDistance) AS hist1MCorr,
  CORR(trip_miles, histThreeMonth_tripDistance) AS hist3MCorr
FROM
  buffer2

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.25s/rows]


histCorr  hist1WCorr  hist1MCorr  hist3MCorr
0  0.924252    0.908758    0.917781     0.92163

The high correlations between the original field `trip_miles` & the four features engineered from historical data `historical_tripDistance`, `histOneWeek_tripDistance`, `histOneMonth_tripDistance` & `histThreeMonth_tripDistance` check the hypothesis and the usage of such features for the modelling part.

### 6-7. pickup_census_tract & dropoff_census_tract

Those fields are important data known at pickup time, but incomplete since the instructions mention that for privacy, those fields may not be shown for some trips.

Let's check the fields.

In [13]:
%%bigquery
-- Extract some basic stats about pickup_census_tract & dropoff_census_tract

SELECT
  COUNTIF(pickup_census_tract IS NOT NULL) AS known_pickups,
  COUNTIF(pickup_census_tract IS NULL) AS unknown_pickups,
  COUNTIF(dropoff_census_tract IS NOT NULL) AS known_dropoffs,
  COUNTIF(dropoff_census_tract IS NULL) AS unknown_dropoffs
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.16s/rows]


known_pickups  unknown_pickups  known_dropoffs  unknown_dropoffs
0      129097545         72004672       128166386          72935831

There are 34% to 35% of unknown pickup or dropoff census tracts.

In [14]:
%%bigquery
-- Extract some basic stats about pickup_census_tract & dropoff_census_tract

SELECT
  COUNT(unique_key) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.21s/rows]


number_trips
0     127244447

There are 65% of trip records with both known pickup & dropoff census tracts.

From the following documentation: https://www.lib.uchicago.edu/e/collections/maps/censusinfo.html, Chicago is divided into 866 census tracts.

For a total area of 606 km2, the average area of one census tract is about 0.7 km2, which can be represented as a square of length __0.5 mile - which imprecision in mapping should cost ~1 USD__.
<br>Let's consider such imprecision acceptable.

Such census tract fields are good candidate to refer to the pickup & dropoff locations.

Let's still take a look at the cardinalities of the census tracts in our dataset.

In [15]:
%%bigquery
-- Extract some basic stats about pickup_census_tract & dropoff_census_tract

SELECT
  COUNT(DISTINCT pickup_census_tract) AS pickup_unique_census_tracts,
  COUNT(DISTINCT dropoff_census_tract) AS dropoff_unique_census_tracts
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.22s/rows]


pickup_unique_census_tracts  dropoff_unique_census_tracts
0                         1291                          1310

There are discrepancies between the official census tract documentation and the census tracts referenced in our dataset.

Actually, filtering based on the official census tracts in Chicago could be a good solution only if we had in hands the official list and a mapping with the census tract ids in our data set.

Let's calculate the cardinalities with some filters:
- `trip_start_timestamp >= '2016-01-01 00:00:00 UTC'` since such rates are applied since January, 1st, 2016;
- `fare > 3.25` since the base fare is USD 3.25;
- `trip_seconds > 0`;
- `trip_seconds < 3800`;
- `trip_miles > 0`;
- `trip_miles / (trip_seconds / 3600) <= 70` since the max speed authorized in Illinois is 70 mph (cf. https://en.wikipedia.org/wiki/Speed_limits_in_the_United_States_by_jurisdiction);
- `pickup_census_tract IS NOT NULL`;
- `dropoff_census_tract IS NOT NULL`.

In [16]:
%%bigquery
-- Extract some basic stats about pickup_census_tract & dropoff_census_tract

SELECT
  COUNT(DISTINCT pickup_census_tract) AS pickup_unique_census_tracts,
  COUNT(DISTINCT dropoff_census_tract) AS dropoff_unique_census_tracts
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
  AND fare > 3.25
  AND trip_seconds > 0 AND trip_seconds < 3800
  AND trip_miles > 0
  AND trip_miles / (trip_seconds / 3600) <= 70
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.27s/rows]


pickup_unique_census_tracts  dropoff_unique_census_tracts
0                          935                          1189

Result is nearer to the reality apparently from the pickup locations (inferior to the 866 official census tracts) but dropoff locations cardinality is still higher, suggesting that there are drops outside of Chicago in our dataset.

### 8-9. pickup_community_area & dropoff_community_area

Those fields are data known at pickup time.

From the following documentation: https://www.lib.uchicago.edu/e/collections/maps/censusinfo.html, Chicago is divided into 77 community areas, which then are coarser than census tracts.

Let's still check the community areas fields.

In [17]:
%%bigquery
-- Extract some basic stats about pickup_community_area & dropoff_community_area

SELECT
  COUNTIF(pickup_community_area IS NOT NULL) AS known_pickups,
  COUNTIF(pickup_community_area IS NULL) AS unknown_pickups,
  COUNTIF(dropoff_community_area IS NOT NULL) AS known_dropoffs,
  COUNTIF(dropoff_community_area IS NULL) AS unknown_dropoffs
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.16s/rows]


known_pickups  unknown_pickups  known_dropoffs  unknown_dropoffs
0      177492003         23610214       173505145          27597072

There are 12% to 14% of unknown pickup or dropoff community areas.

In [18]:
%%bigquery
-- Extract some basic stats about pickup_community_area & dropoff_community_area

SELECT
  COUNT(unique_key) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  pickup_community_area IS NOT NULL
  AND dropoff_community_area IS NOT NULL

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.18s/rows]


number_trips
0     171637287

There are 85% of trip records with both known pickup & dropoff census tracts.

However the community areas are too large to be good indicators of pickup & dropoff locations, since the imprecision of 2.81 miles would represent a cost of USD 6.

### 10. fare

This field will be the target variable to train against and serve at pickup time.

Let's check the field.

In [19]:
%%bigquery
-- Extract basic stats from fare

SELECT
  MIN(fare) AS min_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(10)] AS perc10_fare,
  AVG(fare) AS avg_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(90)] AS perc90_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(95)] AS perc95_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(99)] AS perc99_fare,
  MAX(fare) AS max_fare,
  STDDEV(fare) AS stddev_fare,
  COUNTIF(fare IS NULL) AS missing_values
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.07s/rows]


min_fare  perc10_fare   avg_fare  perc90_fare  perc95_fare  perc99_fare  \
0       0.0         4.75  13.367688         33.5        41.65        51.75   

   max_fare  stddev_fare  missing_values  
0   9999.99    46.078068            6280

This field has a lot of discrepancies: from USD 0 trips to most expensive trips of USD 10,000, which is a bit off.
<br>Furthermore, the field has 3,943 missing values, which can be negligible.

Let's apply some filtering.

Based on the fare rates documentation: https://checkertaxichicago.com/rates-table/, let's apply some filters as previously:
- `trip_start_timestamp >= '2016-01-01 00:00:00 UTC'` since such rates are applied since January, 1st, 2016;
- `fare > 3.25` since the base fare is USD 3.25;
- `trip_seconds >= 36` since the fare is raised every 36 seconds;
- `trip_seconds < 3800`;
- `trip_miles > 0`;
- `trip_miles / (trip_seconds / 3600) <= 70` since the max speed authorized in Illinois is 70 mph (cf. https://en.wikipedia.org/wiki/Speed_limits_in_the_United_States_by_jurisdiction);
- `pickup_census_tract IS NOT NULL`;
- `dropoff_census_tract IS NOT NULL`.

In [20]:
%%bigquery
-- Extract basic stats from fare

SELECT
  MIN(fare) AS min_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(10)] AS perc10_fare,
  AVG(fare) AS avg_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(90)] AS perc90_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(95)] AS perc95_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(99)] AS perc99_fare,
  MAX(fare) AS max_fare,
  STDDEV(fare) AS stddev_fare,
  COUNT(trip_miles) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
  AND fare > 3.25
  AND trip_seconds > 0 AND trip_seconds < 3800
  AND trip_miles > 0
  AND trip_miles / (trip_seconds / 3600) <= 70
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.33s/rows]


min_fare  perc10_fare  avg_fare  perc90_fare  perc95_fare  perc99_fare  \
0      3.26          5.0  12.57815        36.25        44.25         48.0   

   max_fare  stddev_fare  number_trips  
0   9649.45    26.870607      55840046

There are still very large trip durations after the current filtering.
<br>It would be safe to remove the last top 1% which might represent some outliers and fix an upper threshold of __USD 50__ for `fare`.

### 11-12-13. tips, tolls, extras & trip_total

The fields `tips`, `tolls` & `extras` are not known at pickup time.

`tips` & `extras` are related to the customer experience, which might be predicted from the current features above, including the `taxi_id`; but this is definitely beyond the scope of our machine learning framing task.

`tolls` is related to the route taken. To some extent, it would also be possible to predict it with some accuracy from the current features above; however it also is beyond the scope of our machine learning framing task.

As `trip_total` includes `tips`, `tolls` & `extras`, we will also ignore such field.


### 14-15. payment_type & company

Those fields should also not have any importance for our machine learning use case predicting the fares.

Otherwise it would mean that somehow the fares are biased and either there is some surcharge based on the payment mode or some companies may present some irregularities in their driving policies.
<br>Finding such anomalies would result in a specific machine learning use case and is beyond the purpose of our demonstration.
<br>Then let's drop those fields and also exclude them from any exploration.

### 16-17-18-19-20-21. pickup_latitude, pickup_longitude, pickup_location, dropoff_latitude, dropoff_longitude & dropoff_location

Those fields seem to be redundant information from the census tracts and may not bring a lot of additional information.

However census tracts are categorical features and let's engineer from the longitudes & latitudes some numerical features.

Let's check first if such longitudes & latitudes are aligned with their related census tracts

In [21]:
%%bigquery
-- Extract some basic stats about pickup_census_tract & dropoff_census_tract

SELECT
  COUNT(unique_key) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL
  AND pickup_longitude IS NOT NULL
  AND dropoff_longitude IS NOT NULL
  AND pickup_latitude IS NOT NULL
  AND dropoff_latitude IS NOT NULL

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.16s/rows]


number_trips
0     127001203

There are 65% of trip records with both known pickup & dropoff census tracts, both also filled in longitudes & latitudes.

In [22]:
%%bigquery
-- Engineer some features out of pickup_latitude, pickup_longitude, dropoff_latitude & dropoff_longitude

SELECT
  pickup_census_tract,
  dropoff_census_tract,
  SQRT(POW((pickup_longitude - dropoff_longitude),2) + POW(( pickup_latitude - dropoff_latitude), 2)) as rawDistance,     -- Euclidean distance between pickup and drop off
  SQRT(POW((pickup_longitude - dropoff_longitude),2)) as rawLongitude,     -- Euclidean distance between pickup and drop off in longitude
  SQRT(POW((pickup_latitude - dropoff_latitude), 2)) as rawLatitude     -- Euclidean distance between pickup and drop off in latitude
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL
  AND pickup_longitude IS NOT NULL
  AND dropoff_longitude IS NOT NULL
  AND pickup_latitude IS NOT NULL
  AND dropoff_latitude IS NOT NULL
    
  AND fare > 3.25 AND fare < 50
  AND trip_seconds > 0 AND trip_seconds < 3800
  AND trip_miles > 0
  AND trip_miles / (trip_seconds / 3600) <= 70

LIMIT 10

Downloading: 100%|██████████| 10/10 [00:01<00:00,  9.45rows/s]


pickup_census_tract  dropoff_census_tract  rawDistance  rawLongitude  \
0          17031281900           17031062300     0.065143      0.018794   
1          17031062000           17031062300     0.014396      0.014365   
2          17031980100           17031062300     0.179525      0.089491   
3          17031081403           17031062300     0.066210      0.042575   
4          17031081401           17031062300     0.062551      0.041733   
5          17031081300           17031062300     0.059409      0.040681   
6          17031081300           17031062300     0.059409      0.040681   
7          17031081202           17031062300     0.052483      0.035298   
8          17031081201           17031062300     0.055184      0.035233   
9          17031081201           17031062300     0.055184      0.035233   

   rawLatitude  
0     0.062373  
1     0.000949  
2     0.155630  
3     0.050706  
4     0.046595  
5     0.043296  
6     0.043296  
7     0.038840  
8     0.042472  
9     0.042472

We will use the query above as part of __feature engineering__.

Let's now check the hypothesis that the feature `rawDistance` engineered above is informative about the original field `trip_miles`.

In [23]:
%%bigquery
-- Compute the correlations between trip_miles and rawDistance

SELECT
  CORR(trip_miles, rawDistance) AS dist_corr
FROM (

SELECT
  trip_miles,
  SQRT(POW((pickup_longitude - dropoff_longitude),2) + POW(( pickup_latitude - dropoff_latitude), 2)) as rawDistance     -- Euclidean distance between pickup and drop off
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL
  AND pickup_longitude IS NOT NULL
  AND dropoff_longitude IS NOT NULL
  AND pickup_latitude IS NOT NULL
  AND dropoff_latitude IS NOT NULL
    
  AND fare > 3.25 AND fare < 50
  AND trip_seconds > 0 AND trip_seconds < 3800
  AND trip_miles > 0
  AND trip_miles / (trip_seconds / 3600) <= 70

)

Downloading: 100%|██████████| 1/1 [00:01<00:00,  1.10s/rows]


dist_corr
0   0.920568

The high correlation between the original field `trip_miles` & `rawDistance` checks the hypothesis and the usage of such feature for the modelling part.

## Full SQL query

Here is the SQL query to extract the relevant variables, the features to engineer with the right filters as mentioned above.

The results are saved in a specific table.

In [25]:
%%bigquery
-- Create a ML-ready dataset with relevant features and target variable

CREATE OR REPLACE TABLE `aliz-ml-spec-2022-submission.demo1.Demo1_MLdataset` AS (

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
),
    
buffer1 AS (
  SELECT
    unique_key as trip_id,
    
    fare,
    3.25 + trip_seconds / 36 * 0.25 + trip_miles * 2.25 AS primary_fare,
    
    trip_start_timestamp,
    EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
    EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
    daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
    EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
    EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute,
    trip_seconds,
    trip_miles,
    DATE_DIFF(PARSE_DATE('%Y-%m-%d',  FORMAT_TIMESTAMP('%Y-%m-%d', trip_start_timestamp)), '2016-01-01', DAY) refDate,
    pickup_census_tract,
    dropoff_census_tract,
    
    SQRT(POW((pickup_longitude - dropoff_longitude),2) + POW(( pickup_latitude - dropoff_latitude), 2)) as rawDistance,     -- Euclidean distance between pickup and drop off
    SQRT(POW((pickup_longitude - dropoff_longitude),2)) as rawLongitude,     -- Euclidean distance between pickup and drop off in longitude
    SQRT(POW((pickup_latitude - dropoff_latitude), 2)) as rawLatitude     -- Euclidean distance between pickup and drop off in latitude
    
  FROM
    `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames
  WHERE
    trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    
    AND pickup_census_tract IS NOT NULL
    AND dropoff_census_tract IS NOT NULL
    AND pickup_longitude IS NOT NULL
    AND dropoff_longitude IS NOT NULL
    AND pickup_latitude IS NOT NULL
    AND dropoff_latitude IS NOT NULL
    
    AND fare > 3.25 AND fare < 50
    AND trip_seconds > 0 AND trip_seconds < 3800
    AND trip_miles > 0
    AND trip_miles / (trip_seconds / 3600) <= 70
)

SELECT
  * EXCEPT (refDate),
  
  AVG(trip_seconds)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDuration,
  
  AVG(trip_miles)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDistance
FROM
  buffer1

)

Query complete after 0.00s: 100%|██████████| 6/6 [00:00<00:00, 2816.23query/s]                        


Empty DataFrame
Columns: []
Index: []